include/ringmesh/basic/frame.h

```diff
@@ -0,0 +1,188 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *     http://www.ring-team.org
+ *
+ *     RING Project
+ *     Ecole Nationale Superieure de Geologie - GeoRessources
+ *     2 Rue du Doyen Marcel Roubault - TSA 70605
+ *     54518 VANDOEUVRE-LES-NANCY
+ *     FRANCE
+ */
+
+#pragma once
+
+#include <array>
+
+#include <ringmesh/basic/common.h>
+
+/*!
+ * @file Basic classes for representing frames
+ * @author Pierre Anquez
+ */
+
+namespace RINGMesh
+{
+    namespace Geometry
+    {
+        struct Plane;
+        FORWARD_DECLARATION_DIMENSION_STRUCT( Line );
+        ALIAS_2D( Line );
+    }
+}
+
+namespace RINGMesh
+{
+    template < index_t DIMENSION >
+    class RINGMESH_API FrameBase
+    {
+    public:
+        virtual ~FrameBase() = default;
+
+        const vecn< DIMENSION >& operator[]( index_t coord ) const
+        {
+            ringmesh_assert( coord < DIMENSION );
+            return axis_[coord];
+        }
+
+        vecn< DIMENSION >& operator[]( index_t coord )
+        {
+            ringmesh_assert( coord < DIMENSION );
+            return axis_[coord];
+        }
+
+    protected:
+        // Default Frame aligned on space axis
+        FrameBase()
+        {
+            for( auto coord : RINGMesh::range( DIMENSION ) )
+            {
+                axis_[coord][coord] = 1.;
+            }
+        }
+
+    private:
+        std::array< vecn< DIMENSION >, DIMENSION > axis_;
+    };
+    ALIAS_2D_AND_3D( FrameBase );
+
+    template < index_t DIMENSION >
+    class RINGMESH_API Frame : public FrameBase< DIMENSION >
+    {
+    public:
+        Frame() = default;
+    };
+    ALIAS_2D_AND_3D( Frame );
+
+    template <>
+    class RINGMESH_API Frame< 2 > : public FrameBase< 2 >
+    {
+    public:
+        Frame() = default;
+
+        Frame( vec2 x, vec2 y )
+        {
+            ( *this )[0] = std::move( x );
+            ( *this )[1] = std::move( y );
+        }
+    };
+
+    template <>
+    class RINGMESH_API Frame< 3 > : public FrameBase< 3 >
+    {
+    public:
+        Frame() = default;
+
+        Frame( vec3 x, vec3 y, vec3 z )
+        {
+            ( *this )[0] = std::move( x );
+            ( *this )[1] = std::move( y );
+            ( *this )[2] = std::move( z );
+        }
+    };
+
+    template < index_t DIMENSION >
+    class RINGMESH_API ReferenceFrame : public Frame< DIMENSION >
+    {
+    public:
+        ReferenceFrame() = default;
+
+        ReferenceFrame(
+            vecn< DIMENSION > frame_origin, Frame< DIMENSION > frame )
+            : Frame< DIMENSION >( std::move( frame ) ),
+              origin_( std::move( frame_origin ) )
+        {
+        }
+
+        const vecn< DIMENSION >& origin() const
+        {
+            return origin_;
+        }
+
+        vecn< DIMENSION >& origin()
+        {
+            return origin_;
+        }
+
+    private:
```

Comment:
What is the advantage of these 2 functions (getter/setter) instead of a public data member?

Reply:
Keeping the same design for all the Frames. The FrameBase is full private, I do not see why we should open the inherited classes.

Reply:
I think we can choose between this design or using struc and removing all the getters.

Reply:
Between the two solutions you propose I prefer this one.